Corrections in `.normalize_basis_at_p` and `.maximal_order()` of `quaternion_algebra.py`

S17A05 opened the pull request:
- Corrected in `.normalize_basis_at_p` the wrong assignments of `f0, f1`: The idea is to replace `f0` by `f0 + f1` and `f1` by `f0` _simultaneously_ - the original code, however, first replaced `f0` by `f0 + f1` and then copied the value to `f1`, hence causing a decrease in dimension
- Corrected an index error just before the recursive call of `normalize_basis_at_p` in the off-diagnonal case for `p = 2` (the `else`-block)
- In `.maximal_order()` the intermediate basis `e_n` might not define an order anymore (as seen in the comments, this was known to the author(s) of the method) - therefore we need to manually compute the discriminant in the loop instead of relying on the `.discriminant()`-method of quaternion orders
- Adapted an example in `.maximal_order()` to use the new method `.is_maximal()`

Fixes #37217 (both parts) and #37417.

Disclaimer:
I am aware of #37239, but since progress on that PR seems to have halted, I took the liberty to fix the issues myself :)

S17A05 committed:
Corrections in `.normalize_basis_at_p` ...

... and additional fix in `.maximal_order()`.

Amend: Added a space in docstring

S17A05 committed:
Added example from #37417

S17A05 committed:
Removed redundant calculations/calls ...

... plus small refactoring of code

mkoeppe commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -766,9 +765,24 @@ def maximal_order(self, take_shortcuts=True):
         if self.base_ring() != QQ:
-            raise NotImplementedError("maximal order only implemented for rational quaternion algebras")
+            raise NotImplementedError("maximal order only implemented for"
+                                        " rational quaternion algebras")
```
```suggestion
            raise NotImplementedError("maximal order only implemented for "
                                      "rational quaternion algebras")
```

mkoeppe commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -892,13 +908,23 @@ def maximal_order(self, take_shortcuts=True):
+                    MM = []
+                    for y in e:
+                        MM.append(x.pair(y))
+                    L.append(MM)
```
This could be rewritten as list comprehensions

mkoeppe commented on src/sage/algebras/quatalg/quaternion_algebra.py:
```diff
@@ -892,13 +908,23 @@ def maximal_order(self, take_shortcuts=True):
+                    for y in e:
+                        MM.append(x.pair(y))
+                    L.append(MM)
+                disc = (MatrixSpace(QQ, 4, 4)(L)).determinant().sqrt()
```
```suggestion
                disc = matrix(QQ, 4, 4, L).determinant().sqrt()
```

S17A05 committed:
Implemented reviewer feedback

- Blank spaces in error warning
- Matrix definition
- List comprehension for discriminant computation, also updated in `.discriminant()` of an order

S17A05 committed:
Merge branch 'develop' into normalize_maximal_fix

mkoeppe reviewed (approved):
LGTM
